racket/collects/xml/private/reader.rkt

```diff
@@ -332,50 +332,39 @@
         (loop)))))
 
 ;; lex-pcdata : Input-port (-> Location) -> Pcdata
-;; deviation - disallow ]]> "for compatibility" with SGML, sec 2.4 XML spec 
+;; deviation - disallow ]]> "for compatibility" with SGML, sec 2.4 XML spec
 (define (lex-pcdata in pos)
-  (let ([start (pos)]
-        [data (let ()
-                (define out (open-output-string))
-                (let loop ()
-                  (let ([next (peek-char-or-special in)])
-                    (cond
-                      [(or (eof-object? next)
-                           (not (char? next))
-                           (eq? next #\&)
-                           (eq? next #\<))
-                       (void)]
-                      [(and (char-whitespace? next) (collapse-whitespace))
-                       (skip-space in)
-                       (write-char #\space out)
-                       (loop)]
-                      [else
-                       (write-char next out)
-                       (void (read-char in))
-                       (loop)])))
-                (get-output-string out))])
-    (make-pcdata start (pos) data)))
+  (define start (pos))
+  (define data
+    (let loop ([chars null])
+      (let ([next (peek-char-or-special in)])
+        (cond
+          [(or (eof-object? next)
+               (not (char? next))
+               (eq? next #\&)
+               (eq? next #\<))
+           (apply string (reverse chars))]
+          [(and (char-whitespace? next) (collapse-whitespace))
+           (skip-space in)
+           (loop (cons #\space chars))]
+          [else
+           (loop (cons (read-char in) chars))]))))
+  (make-pcdata start (pos) data))
 
 ;; lex-name : Input-port (-> Location) -> Symbol
 (define (lex-name in pos)
-  (let ([c (non-eof read-char-or-special in pos)])
-    (unless (name-start? c)
-      (lex-error in pos "expected name, received ~e" c))
-    (string->symbol
-     (let ()
-       (define out (open-output-string))
-       (write-char c out)
-       (let lex-rest ()
-         (let ([c (non-eof peek-char-or-special in pos)])
-           (cond
-             [(eq? c 'special)
-              (lex-error in pos "names cannot contain non-text values")]
-             [(name-char? c)
-              (write-char c out)
-              (void (read-char in))
-              (lex-rest)]
-             [else (void)])))
-       (get-output-string out)))))
+  (define c (non-eof read-char-or-special in pos))
+  (unless (name-start? c)
+    (lex-error in pos "expected name, received ~e" c))
+  (let loop ([chars (list c)])
+    (define c (non-eof peek-char-or-special in pos))
+    (cond
+      [(eq? c 'special)
+       (lex-error in pos "names cannot contain non-text values")]
+      [(name-char? c)
+       (loop (cons (read-char in) chars))]
```

Comment:
[non-blocking: I'm not well-versed in this code] This looks like it throws away c, but the original code threw away the next read; is c from a "peek", which I suppose would make these are equivalent?

Reply:
>  is c from a "peek", which I suppose would make these are equivalent?

Yes. The previous code peeked c then threw away the read. This code peeks then uses the read. The behavior is equivalent, but this version is shorter.